(Custom)SelectControl: Refresh and refactor chevron

mirka opened the pull request:
Follow-up to #42460

## What?

Refactors the `chevronDown` icon in SelectControl/CustomSelectControl so the implementation is clean and the visual is consistent.

### Visual changes

- Paddings are now dynamic based on the size variant.

@jasmussen Are we happy with the size of the icon? And the spacing before and after it? Let me know if you have tweaks.

## Why?

We now have a clean way to implement the chevron (`suffix` prop & suffix wrapper component). It removes hacks and maintains visual consistency across components.

## How?

Refactor the implementation to use the `suffix` prop & suffix wrapper component.

## Testing Instructions

1. `npm run storybook:dev`
2. See the stories for SelectControl and CustomSelectControl. Check the size variants.

✅ Works in RTL
✅ Width behavior is maintained for CustomSelectControl. (130px wide by default, and grows when the label is long)

## Screenshots or screencast <!-- if applicable -->

Here's a video of all the size variants (24, 30, 36, 40):

https://user-images.githubusercontent.com/555336/182780159-c407e2a2-56bc-4d5a-9348-78e51838bb7a.mp4

mirka committed:
(Custom)SelectControl: Refresh and refactor chevron

mirka commented on packages/components/src/custom-select-control/styles.ts:
```diff
@@ -0,0 +1,28 @@
+const backCompatMinWidth = ( props: BackCompatMinWidthProps ) =>
+	! props.__nextUnconstrainedWidth
+		? css`
+				${ InputControlContainer } {
+					min-width: 130px;
+				}
```
Unfortunate that we need an override, but this is only for temporary back compat, and I don't expect us needing to do this often.

ciampo replied:
Apologies in advance because I may have forgotten this detail, but — is the plan here to follow our ["deprecating styles" strategy](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#deprecating-styles)?

mirka replied:
Correct! I'm working on this as part of the follow-ups for #42460, and will post the PR soon.

mirka committed:
Add changelog

jasmussen commented:
Nice! This looks pretty good to me. Here's the reference I'm looking at for comparison:

<img width="305" alt="Screenshot 2022-08-04 at 13 17 35" src="https://user-images.githubusercontent.com/1204802/182834489-f310f065-9b4d-435a-b377-f59df6718e5e.png">

As I mentioned on another thread, I'm refreshing my memory and coming back to the conversation about size defaults — this defaults to 40px correct?

mirka commented:
> this defaults to 40px correct?

No, the default still remains 30px. I'm inching closer to getting 40px variants enabled in the sidebar though. I'm planning on flipping the switch beginning with the Typography panel in the Global Styles sidebar, since that's pretty contained.

ciampo commented on packages/components/src/select-control/styles/select-control-styles.ts:
```diff
@@ -138,17 +139,6 @@ export const Select = styled.select< SelectProps >`
-	svg {
-		display: block;
-	}
+	margin-inline-end: ${ space( -1 ) };
```
Interesting that this is necessary even when using `InputControlSuffixWrapper`

mirka replied:
Yeah, this one's less about the outer padding wrapper, and more about wanting to offset the whitespace in the SVG viewbox, if that makes sense.

ciampo commented on packages/components/src/custom-select-control/style.scss:
```diff
@@ -12,18 +12,6 @@
 	text-align: left;
 	outline: 0; // focus ring is handled elsewhere
-
-	&:not(.is-next-unconstrained-width) {
-		min-width: 130px;
-	}
```
Given this deletion, is assigning this classname still necessary?

https://github.com/WordPress/gutenberg/blob/4137f40fc9c9b2831a1579e55d5e6825595532e5/packages/components/src/custom-select-control/index.js#L172-L175

mirka replied:
Good catch 🙏

ciampo commented on packages/components/src/custom-select-control/styles.ts:
```diff
@@ -0,0 +1,28 @@
+};
+
+const backCompatMinWidth = ( props: BackCompatMinWidthProps ) =>
+	! props.__nextUnconstrainedWidth
```
It looks like `InputBaseWithBackCompatMinWidth` never receives the `__nextUnconstrainedWidth` prop in `packages/components/src/custom-select-control/index.js`

mirka replied:
Lol, I like how this still accidentally works even without the correct prop 😂 (Also how do people code without end-to-end typescript checking...)

ciampo reviewed (commented):
Thank you for working on this!

Currently, when I try to click on the chevron icon, the dropdowns (both in `SelectControl` and `CustomSelectControl`) don't react to the pointer events (clicking on the rest of the input works)

Also, the layout of `SelectControl` looks a bit broken when switching to "Wordpress (common/forms)" in the global CSS switcher

![Screenshot 2022-08-08 at 19 08 58](https://user-images.githubusercontent.com/1083581/183474543-fb119821-54f8-48c1-8b1d-0088bb8c18d5.png)

mirka committed:
Remove unneeded classname

mirka committed:
Fix wrong prop passing

mirka committed:
Restore clickability

mirka committed:
Merge branch 'trunk' into clean-select-chevron

# Conflicts:
#	packages/components/CHANGELOG.md
#	packages/components/src/custom-select-control/index.js

mirka committed:
Fix max-width problem in forms.css

mirka committed:
Normalize to cursor: pointer

mirka commented:
> Currently, when I try to click on the chevron icon, the dropdowns (both in `SelectControl` and `CustomSelectControl`) don't react to the pointer events (clicking on the rest of the input works)

Great one, thank you. Fixed in f5f5639.

> Also, the layout of `SelectControl` looks a bit broken when switching to "Wordpress (common/forms)" in the global CSS switcher

Congrats, this observation caught two pre-existing forms.css problems:

- forms.css was setting `max-width: 25rem` on `select`s (ab0143e)
- forms.css was setting `cursor: pointer` on `select`s, so SelectControl got a pointer but CustomSelectControl didn't (dc8fe6c)

Thanks for the great review @ciampo! All feedback is now addressed.

ciampo commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -147,12 +148,15 @@ export default function CustomSelectControl( {
 			) }
-			<InputBase
+			<InputBaseWithBackCompatMinWidth
+				__nextUnconstrainedWidth={ __nextUnconstrainedWidth }
```
I believe there is also a need for `__next36pxDefaultSize` to be passed to `InputBaseWithBackCompatMinWidth` , since the underlying [`InputBase` accepts it as a prop](https://github.com/WordPress/gutenberg/blob/f4dde1cca860fb74bf8b7a2644175bb06bae3102/packages/components/src/input-control/input-base.tsx#L57) and uses it to set the correct prefix/suffix padding?

ciampo commented on packages/components/src/select-control/styles/select-control-styles.ts:
```diff
@@ -8,7 +8,9 @@ import styled from '@emotion/styled';
 import { COLORS, rtl } from '../../utils';
+import { space } from '../../ui/utils/space';
 import type { SelectControlProps } from '../types';
+import InputControlSuffixWrapper from '../..//input-control/input-suffix-wrapper';
```
Double slash: 

```suggestion
import InputControlSuffixWrapper from '../../input-control/input-suffix-wrapper';
```

ciampo commented on packages/components/src/select-control/styles/select-control-styles.ts:
```diff
@@ -138,17 +146,14 @@ export const Select = styled.select< SelectProps >`
+	InputControlSuffixWrapper
+)`
+	position: absolute;
+	right: 0;
```
I believe these styles are missing RTL support:

```suggestion
	${ rtl( { right: 0 } ) }
```

ciampo reviewed (commented):
Looking better! 

> Congrats, this observation caught two pre-existing forms.css problems:

![jackpot](https://user-images.githubusercontent.com/1083581/184088448-5d52f59b-4b17-4d2d-b5c1-9540e7e3a3f6.gif)

mirka committed:
Restore `__next36DefaultSize` prop

mirka committed:
Add RTL support

mirka committed:
Remove extra slash

mirka commented:
@ciampo What would we do without you, Marco. Thanks for the 👀 !

ciampo reviewed (approved):
LGTM 🚀 

This was a tricky one to get right, great job!

It's gratifying to see the steady little steps towards making our components more coherent ⭐
